### ISSUE TITLE
Sample Resource System Design

### ISSUE THREAD

**Issue by ealmloff** [KerfuffleV2/llm-samplers#8]
> Definitely! This actually ties into something else: the poor state of the sampler resource system. > You'd really want a good way to store your parser state (so you can avoid reparsing everything repeatedly like you said).
> 
> I started trying to revamp it, but ran into issues and kind of gave up. I'll write you a better response in a few days (please ping me if you don't hear back). Even before you said that, I was already thinking with your Dioxus experience probably would have some good ideas for how to solve that (assuming you'd be interested in helping with/contributing to that kind of thing).

@KerfuffleV2 I'm not entirely sure what your goals are with the sampler resources system. There are two main approaches I have seen within different projects:
1) Trait/tuple approach: Axum uses this approach with the [FromRef](https://docs.rs/axum/latest/axum/extract/trait.FromRef.html). It makes it so that middleware (or samplers in a chain) can check at compile time if the resource has been inserted. It allow samplers to check if dependancies have been added before the chain to verify that the sampling order is correct
2) AnyMap approach: Dioxus uses a Rc<RefCell<[anymap](https://crates.io/crates/anymap)>> to store context state. llm-samplers could use a similar approach to allow any state to be inserted into the map. There could be functions that insert the correct type for common resources like the previous tokens. This could be more flexible because the types are erased so you don't need to deal with any generics, but it cannot verify the sampling order at compile time

**Comment by KerfuffleV2** [comment]
Really sorry I didn't follow up on that. Approach 2 is the one I was looking at:

>  but it cannot verify the sampling order at compile time

I think we might be talking about different things though. By "resources" I mean the resources the sampler needs access to. For example the logits (RW access), the RNG (RW access), the last tokens (RO access). Right now those resources are hardcoded and if you want to give a sampler access to some other kind of resource (just for example, a classifier-free guidance sampler would need access to the CFG logits and maybe some other params from it).

Some samplers may need RW access to multiple types of resource at the same time. For example Mirostat manipulates/filters the logits and also uses RNG to pick a token (it calls the rand distrib sampler internally).

That's already getting a little complicated because we're talking about needing mutable access to multiple things at the same time, but if it's at all possible I also really want to avoid having to copy data around. For example, with long context models that support 128K context size, that last tokens resource could have 100K+ items in it. There are also models with very large vocab. LLaMA models have a pretty small vocab (you probably already know this) at only 32K entries, but there are some models with 250K+ vocab. So that means there will be 250K logits items before filtering.

So what I really want to do is access a reference for those things instead of making the user actually copy them for sampler use, but that means mutable access to multiple slices at the same time. And that's basically where I ran into trouble and didn't manage to make further progress... So I got distracted and wandered off.

I stuck what I had so far in a branch: https://github.com/KerfuffleV2/llm-samplers/tree/resource-experiment - Note, this is incomplete and doesn't even compile currently. 

`Any` doesn't normally support lifetimes but from what I found out, doing this is actually legal: https://github.com/KerfuffleV2/llm-samplers/blob/b670b508f2d23ee605a980fecc69ad533702b3cd/src/resource.rs#L62-L89

And we could say that whatever resources get type-erased have to have a lifetime at least however long. The problem is trying to access multiple mutable ones simultaneously, like the RNG state + logits. *edit*: I'm not really concerned with compile-time verification that the resources exist. I think it's fine to fail at runtime if a sampler tries to run but the appropriate resources weren't provided. So they can just try to pull something out of the map or whatever, and if it's not there, return an `Err`.

**Comment by ealmloff** [comment]
That does seem sound. https://docs.rs/better_any/latest/better_any/index.html might be useful as a more general approach to Any with lifetimes

**Comment by ealmloff** [comment]
To get multiple references to elements in the hashmap, you can use https://rust-lang.github.io/hashbrown/hashbrown/struct.HashMap.html#method.get_each_mut in the `hashbrown` library (this is the library that backs `std`'s hashmap)

**Comment by KerfuffleV2** [comment]
> https://docs.rs/better_any/latest/better_any/index.html might be useful as a more general approach to Any with lifetimes

Thanks! I didn't find that one when I was looking around. I don't mind adding relatively lightweight dependencies like that.

> To get multiple references to elements in the hashmap, you can use https://rust-lang.github.io/hashbrown/hashbrown/struct.HashMap.html#method.get_each_mut 

This looks really interesting. But:

"*This method is available only if the `nightly` feature is enabled.*"

**Comment by ealmloff** [comment]
> This looks really interesting. But:
> 
> "This method is available only if the nightly feature is enabled."


I think I linked to the wrong docs. That is the version that rust uses. I think that nightly feature lines up with std's version of nightly. The crates.io version of the crate has that feature on stable: https://docs.rs/hashbrown/0.14.2/hashbrown/hash_map/struct.HashMap.html#method.get_many_mut

**Comment by KerfuffleV2** [comment]
Oh, thanks! Sorry, I should have done a little more digging there. I've started work on v0.0.7 in #9, however that one probably won't have the resource system overhaul.

It involves one pretty large change, removing type variables from the samplers and just always using `u32` for token ids and `f32` for logits. This massively, massively simplifies basically everything and also makes the interface easier to use.

I also found a weird bug in the Mirostat2 sampler where it was sampling `rand_distrib` twice. That probably won't have a huge effect on performance but I believe you were using Mirostat2 in your chain from the other issue so maybe it'll be a tiny bit faster now.
